### ISSUE TITLE
[usbdev] Check deadlines for responding to back-to-back packets

### ISSUE THREAD

**Issue by a-will** [lowRISC/opentitan#10063]
This is mostly to surface potentially hazardous assumptions about the speed of the interrupt service routine. There are some cases of back-to-back packets that depend on the ISR being fast enough to respond to the first packet and modify registers in usbdev before it's time to send the appropriate response to the second packet.

Since full-speed USB isn't particularly fast, there might not be a significant danger. However, it would be prudent to look deeper at the use case and timing, then see if design changes are needed.

- Back-to-back IN transactions to the same endpoint
   - Only one buffer can be queued at once for a given endpoint. The interrupt for the first packet's completion would fire upon delivering the ACK to the host, and the ISR would need to enqueue the next buffer before the next IN token completes.
   - Deadline: ~2.8 us. (2 + 8 + 8 + 16) = 34 bit times at 12 MHz. Interpacket gap + SYNC + PID + the rest of the token packet.
- Back-to-back OUT transactions, where a STALL or NAK handshake is desired for the second packet
   - The buffer pool for RX is shared among all the endpoints, and there is no control to automatically set a STALL or NAK response for a given endpoint after the next transaction. The ISR must respond after the first transaction is received to set up a STALL or NAK response before the handshake period of the second transaction.
   - Deadline: ~6.58 us. (37 + 2 + 8 + 8 + 16 + 8) = 79 bit times at 12 MHz. Token packet from above (with the EOP added) + Interpacket gap + SYNC + PID + CRC16 + 1 byte of data.

**Comment by GregAC** [comment]
@alees24 is this something that needs fixing given the comments in https://github.com/lowRISC/opentitan/issues/17239 that we'd be 'extremely unlikely' to meet these deadlines it is something we have to address?

**Comment by alees24** [comment]
As far as I'm aware missing these deadlines does not induce functional failures, but it is a potential issue as far as throughput is concerned, as noted in #17239. The concern is that if the host controller receives a NAK indicating that there is no new buffer available for INput, it may not retry the endpoint in question until the next (1ms) bus frame occurs, which could substantially throttle the bandwidth in the case of not managing back-to-back IN traffic.

Back-to-back OUT transactions should not be an issue w.r.t. performance and to my knowledge there is no case where failing to activate STALL/NAK response promptly leads to a functional failure.

**Comment by alees24** [comment]
I've locally extended the software such that it can provide new IN packets without incurring the delay of first writing the data into the usbdev buffer memory; essentially (optional) double-buffering per endpoint in an effort to reduce the latency and increase the throughput, and I'm seeing a measured 5.6us latency in Verilator chip sim at 48MHz (synchronous Ibex and usbdev), which I estimate could drop to 3.4us. That's the time from the pkt_sent signal assertion to the configin write with the new packet description.

This is still polling-based code but modified with an eye to avoiding unnecessary work, subject to the constraint that it's still using the DIF for all hardware access within that interval, and so incurs its overheads. I'd imagine any production interrupt handler would be doing a comparable amount of work to the code that's executing at the moment, minus some of the DIF overhead.

@a-will My observations in Verilator sim (with DPI model) report that we need to be sub 3.6us to achieve back-to-back transfer and stop usbdev from NAKing IN requests, so the same ball park as your estimate; kudos!

If we really had to drop the interrupt latency low enough to achieve back-to-back transfers, I'm confident that it could be achieved by more careful coding and eliminating some of the inefficiencies, but _only in the event that we're not delayed by other interrupt handlers executing_. Since the data we're transferring must be coming from/going to somewhere else, there will surely be other interrupt handlers executing.

From running this code on the FPGA (with a 10MHz Ibex, so there's no way that it's achieving back-to-back transfers) it does seem that the host controller makes some effort at retrying transfers during the same bus frame. I'm seeing an average of 8.9 packet transfers per bus frame from the single endpoint being used, even though a back-to-back IN attempt will have resulted in at least one NAK.

@a-will I'm not sure that I fully understand the (possible) desire for STALL/NAK handshake on a subsequent packet; I understand that the buffers have been supplied up front, and thus modification of the registers is necessary to prevent the packet arriving. Is there some danger of a functional breakage that I've missed, or just that it may complicate the software to have packets arriving and it would be desirable to hold them off?

**Comment by a-will** [comment]
@alees24 STALL is sometimes used as part of a higher-layer protocol atop functions (the type that sets the halt bit). There is some danger of breakage if packet acceptance is meaningful for such functions, for example, where the host incorrectly thinks data was processed.

Without help from the other software-controlled CSRs, usbdev might not have all the up-to-date info for determining how to handle the incoming packet. If the CSRs don't update fast enough, the only dynamic info would be whether a buffer is available locally for the packet in the shared free RX buffer queue. That is not always the full set of conditions.

I'm fuzzy on particular interface classes that need this, though.

**Comment by alees24** [comment]
> estimate 16
> remaining 2023-03-23 8

**Comment by johngt** [comment]
Update Jan-23. This should remain open. May still have impact upon USB classes/protocols. Remains to be determined if it can induce functional failures for particular USBDEV use cases. From @alees24

**Comment by alees24** [comment]
@a-will I had a look through the main class specifications and could not find any reference to use of 'functional stall'

If we're concerned about packets being received inappropriately, do you think there's merit to introducing a 'set_stall_out' register that holds, per endpoint, a single bit which will cause the 'out_stall' bit to be set upon receipt of an OUT packet to the endpoint in question, mimicking the way that 'set_nak_out' presently clears the 'rxenable_out' bit?

The 'set_nak_out' already gives us per-endpoint rate limiting using NAK responses, with packet level granularity, for anything that _must not_ accept packets before it can guarantee to handle them.

This would be a low-impact change as a contingency.

**Comment by a-will** [comment]
@alees24 I think I might've written this issue before I introduced `set_nak_out` (and it may have motivated the latter). If an endpoint is using that mode (for every packet to that endpoint), it is delaying responses to each packet until software can decide whether it is allowed to accept. Then a STALL response doesn't have to be immediate.

By contrast, a `set_stall_out` sounds a bit like we could end up in a similar race. It wouldn't buy software substantially more time to respond, would it?

It's been awhile since I looked at usbdev, so I could be missing things, haha. However, it sounds to me like you pointed out how `set_nak_out` can handle this issue already.
